### ISSUE TITLE
Randomly clearing files in /tmp/

### ISSUE THREAD

**Issue by etherealvisage** [gogs/gogs#5914]
**Describe the bug**
All files in `/tmp/` that the user running gogs can access are cleared regularly.

**Gogs version or commit**

0.11.97.1209

**Git version**

```
$ git version
git version 2.20.1
```

**Operating system**

Debian testing, aarch64

**Database**

sqlite. version information:
```
$ sqlite3 --version
3.27.2 2019-02-25 16:06:06 bd49a8271d650fa89e446b42e513b595a717b9212c91dd384aab871fc1d0alt1
```

**To Reproduce**
1. Run `./gogs web`, then `mkdir /tmp/a_file` as the same user.
2. Wait.
3. Directory will be removed.

**Can you reproduce the bug at https://try.gogs.io?**

No, I don't have access to the filesystem there.

**Expected behavior**

I expected only relevant files in `/tmp/` to be removed by gogs, such as those it creates itself.

**Actual behavior**

Files in `/tmp/` that are writable by the user running gogs are removed.

I see entries like the following in the `auditd` log:
```
type=SYSCALL msg=audit(1580711882.922:259): arch=c00000b7 syscall=35 success=no exit=-2 a0=ffffff9c a1=aaab1df8f490 a2=0 a3=ffffbcf82780 items=1 ppid=4526 pid=1121 auid=0 uid=0 gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0 tty=(none) ses=2 comm="rm" exe="/bin/rm" key=(null)
```
The `ppid` `4526` is the PID of the gogs binary.

**Additional context**

All files in `log/gogs.log` appear normal, with things like:
```
2020/02/03 01:53:14 [ INFO] Gogs 0.11.97.1209
2020/02/03 01:53:14 [ INFO] Cache Service Enabled
2020/02/03 01:53:14 [ INFO] Session Service Enabled
2020/02/03 01:53:14 [ INFO] Git Version: 2.20.1
2020/02/03 01:53:14 [ INFO] SQLite3 Supported
2020/02/03 01:53:14 [ INFO] Run Mode: Production
2020/02/03 01:53:14 [ INFO] SSH server started on 0.0.0.0:22222
2020/02/03 01:53:15 [ INFO] Listen: http://127.0.0.1:9006
```

**Comment by unknwon** [comment]
Thanks for a high quality bug report!

First of all, IIRC, Gogs does not even use the system's `/tmp` directory, it has its own "tmp" subdirectory under the app data path. By default, it is `./data/tmp`:

https://github.com/gogs/gogs/blob/465be9b16ee331c258a5a4fa29921953e64e5266/conf/app.ini#L68

There is only one place trying to delete the "tmp" directory, which is when the Gogs web server starts:

https://github.com/gogs/gogs/blob/465be9b16ee331c258a5a4fa29921953e64e5266/internal/db/repo.go#L147

Can you check what's the value of this **APP_DATA_PATH**?

**Comment by etherealvisage** [comment]
It's empty in the custom configuration path, which would probably be why that's happening, then. :-) I guess the join of the paths `""` and `tmp` is `/tmp`?

```
$ grep APP_DATA_PATH gogs/custom/conf/app.ini
$ 
```

As for _why_ it's like that, I'm not sure. Here's the entire app.ini configuration in case it's relevant (with some paths etc elided):

```
anfil:~/gogs/custom/conf$ cat app.ini 
APP_NAME = Gogs
RUN_USER = ethereal
RUN_MODE = prod

[database]
DB_TYPE  = sqlite3
HOST     = 127.0.0.1:3306
NAME     = gogs
USER     = root
PASSWD   = 
SSL_MODE = disable
PATH     = data/gogs.db

[repository]
ROOT = <elided>

[server]
DOMAIN           = <elided>
HTTP_PORT        = 9006
HTTP_ADDR        = <elided>
ROOT_URL         = <elided>
DISABLE_SSH      = false
SSH_LISTEN_HOST  = 0.0.0.0
SSH_PORT         = 22222
START_SSH_SERVER = true
OFFLINE_MODE     = false

[mailer]
ENABLED = false

[service]
REGISTER_EMAIL_CONFIRM = false
ENABLE_NOTIFY_MAIL     = false
DISABLE_REGISTRATION   = true
ENABLE_CAPTCHA         = false
REQUIRE_SIGNIN_VIEW    = true

[picture]
DISABLE_GRAVATAR        = false
ENABLE_FEDERATED_AVATAR = true

[session]
PROVIDER = file

[log]
MODE      = file
LEVEL     = Info
ROOT_PATH = <elided>

[security]
INSTALL_LOCK = true
SECRET_KEY   = <elided>
```

**Comment by unknwon** [comment]
If the key is not presented in the `custom/conf/app.ini`, it uses the value from the default `conf/app.ini`, which should be "data". And to trigger the "deletion", your Gogs web service has to been restarted. I tested on my mac, with all defaults, a restart won't delete anything in my `/tmp`. So I think it might be something else.

I looked at my admin panel again, I found there is one case that Gogs uses the "/tmp" if not configured. Gogs uses `/tmp/gogs_keytest/<random_string>` to test if a SSH public key is valid, but also seems not the cause of the problem you described because the `os.Remove` will fail on directories (can only remove a file). Ref: https://sourcegraph.com/github.com/gogs/gogs@465be9b16ee331c258a5a4fa29921953e64e5266/-/blob/internal/db/ssh_key.go#L182-200


🤔 Running out of ideas.

**Comment by etherealvisage** [comment]
I believe `os.Remove` will also [remove empty directories](https://golang.org/pkg/os/#Remove)? That seems like it could be it, but there were definitely non-empty directories being removed (I first noticed the problem when my `/tmp/tmux-<uid>` directories were disappearing). Of course, there could be _another_ service running on the machine removing things in `/tmp/`, but as much as I'm a fan of gogs, I'm going to apply occam's razor here :-)

Looking through the `auditd` logs again, I found something ... _interesting_. The file isn't removed by the gogs executable directly, but instead by `/bin/rm`.

```
type=CWD msg=audit(1580711685.006:78): cwd="/tmp/.X19-unix/.rsync/c"
type=PATH msg=audit(1580711685.006:78): item=0 name="/tmp/tmux-1024" inode=1035520 dev=00:1a mode=040755 ouid=1024 ogid=1024 rdev=00:00 nametype=NORMAL cap_fp=0 cap_fi=0 cap_fe=0 cap_fver=0 cap_frootid=0
type=PROCTITLE msg=audit(1580711685.006:78): proctitle=726D002D7266002F746D702F746D75782D31303234
type=SYSCALL msg=audit(1580711685.006:79): arch=c00000b7 syscall=35 success=yes exit=0 a0=ffffff9c a1=aaaac86a9490 a2=200 a3=ffffbee24780 items=2 ppid=4526 pid=32352 auid=0 uid=0 gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0 tty=(none) ses=2 comm="rm" exe="/bin/rm" key=(null)
```
That working directory looked suspicious to me (`/tmp/.X19-unix/.rsync/c`) and lo and behold, [it's a path used by some malware](http://webcache.googleusercontent.com/search?q=cache:0f900O_Cgy0J:https://blog.martdj.nl/2020/01/08/i-got-hacked-no-not-this-blog/&client=firefox-b-1-d&hl=en&gl=us&strip=1&vwsrc=0).

Now the question is how on _earth_ it got onto the machine --- passwords are disabled on all but one account (which has a 64-character random password) and I'm willing to bet a _lot_ that the pubkeys that are allowed in are secure. Given that it _was_ running as a process spawned by gogs, I feel like perhaps there's something wrong with how I set up gogs' SSH server / there's a flaw in gogs' SSH server setup? Any advice you can offer here?

**Comment by unknwon** [comment]
I have no idea what's going at the moment. But I saw you're configured to use builtin SSH server on 22222, so few notes about correct SSH setup off the top of my head:
1. Your clone URL for should looks like `ssh://domain:22222/<owner>/<repo>`.
2. Your `.ssh/authorized_keys` for the user that runs Gogs, should not contain any SSH keys added by Gogs.

**Comment by etherealvisage** [comment]
Alright, I've managed to track it down. Doesn't look like it's gogs' fault --- at some point a piece of cryptominer malware, and it appeared to not expect an aarch64 system, which caused it to exhibit the failure mode of repeatedly removing things in `/tmp/`. I've imaged the drive and will be running what forensics I can to figure out if I can figure out where it came from or why it was running under the user gogs was running under. I personally suspect something to do with the gogs ssh server setup given circumstances, but honestly that's only because I'm at a loss for how it could have arrived in the first place so I'm grasping at straws :-) I'll audit gogs' SSH server code if I get the time.

Sorry to bother you with all this! Thanks again for writing gogs --- current experience notwithstanding, it's been a great tool. :-)
